### DISCUSSION TITLE
Safe to access to a protected member _service_class of falconpy? [YES/SAFE]

### DISCUSSION THREAD

**Discussion by morcef** [CrowdStrike/falconpy#553] [Q&A] [answered]
Hi,

I am building some custom clients that make use of functionality in your service classes. I have an abstract BaseClient which handles some common functionality, such as initializing the relevant service class and handling errors. In my BaseClient I would like to use `__init__` from `ServiceClass`, as well as typing the service each of my clients use for IDE autocompletion/hinting.

The following is a very basic example of what I wish to accomplish:

```
from abc import ABCMeta
from typing import Type

from falconpy import OAuth2, Detects, Incidents
from falconpy._service_class import ServiceClass


class FalconBaseClient(metaclass=ABCMeta):

    service_class: Type = ServiceClass
    service: ServiceClass

    def __init__(self, auth_object: OAuth2):
        self.service = self.service_class(auth_object=auth_object)
        
        if not self.service.authenticated():
            self.some_method_to_handle_error()

    def some_method_to_handle_error(self):
        
        pass


class FalconDetectsClient(FalconBaseClient):

    service_class: Type = Detects
    service: Detects

    def get_detect_ids(self) -> dict:

        return self.service.query_detects()


class FalconIncidentsClient(FalconBaseClient):

    service_class: Type = Incidents
    service: Incidents

    def get_incident_ids(self) -> dict:
        
        return self.service.query_incidents()

```

This code currently works, but my following question is based on the warning from this `from falconpy._service_class import ServiceClass`:
```
Access to a protected member _service_class of a module
```

Question: Why does this module have a leading underscore? Do you foresee any breaking changes happening in this module, so that I cannot rely on it? Would you have any other suggested solutions to accomplish the same?

**Comment by jshcodes** [comment, marked as answer]
This looks a lot like the experiments I've been playing with in the [Caracara alpha source](https://github.com/CrowdStrike/caracara). With my goal being very similar... abstract things like error handling, rate limiting and individual endpoint query maximums. (Along with other cool conjoined functionality leveraging multiple endpoints.) 

Answering your questions in order:
1. ServiceClass was defined as a private only because it was being used as a base class for all Service Classes. My thinking at the time being, "this is an internal base class that we'll use to define all subsequent public classes." (with the (probably wrong) assumption being you'd inherit from the public service class based upon API need.)
2. The ServiceClass base class plays a critical role, there are no plans to change this.  There _has_ been change in this class, (like when we added in `token_fail_reason`) but this is managed very carefully due to the important role this class plays. We unit test heavily to reduce the potential for breaking change impacting your usage, but nothing is ever perfect. If you encounter problems using this class as you've described, or if future change impacts your usage, please post an issue. It is more than likely a bug. To answer the core of your question, we do __not__ have plans to remove or change this base class or it's name. I don't foresee this changing any time soon.
3. Check out what I have in the Caracara source so far.  It's pretty ugly at the moment as I'm still noodling with ideas.  We'd love to collaborate with you on this as I think your approach as described is more elegant.

**Reply by jshcodes** [reply]
Follow up: I have another project that leverages the Caracara package to emulate a RTR console that I've been using as a test bed. I'll get this posted ASAP so you can see some examples of usage.

**Reply by morcef** [reply]
Cool! I'll have a look into that project :)

And thank you for your answers! They were as expected, so I guess I will just live with the warning when knowing this.

**Reply by jshcodes** [reply]
I'm thinking on this one.  There are a couple of solutions, just need to do some discovery to investigate the impacts.

**Reply by jshcodes** [reply]
Hi @morcef!

So I believe I've come up with a solution where I export the base class, `ServiceClass`, in `__init__.py` as part of `__all__`.

I'm still testing this but wanted to give you a chance to check it out and let me know if the fix works for you.

1. Remove the current version of crowdstrike-falconpy.
    ```shell
    python3 -m pip uninstall crowdstrike-falconpy
    ```
2.  Install the development version of the crowdstrike-falconpy-dev package.
    ```shell
    python3 -m pip install crowdstrike-falconpy-dev==1.0.4.dev1
    ```

3. Update your code to reference ServiceClass straight from the package.
    ```python3
    from falconpy import ServiceClass

    generic = ServiceClass(client_id="CLIENT_ID", client_secret="CLIENT_SECRET")
    ```

**Reply by jshcodes** [reply]
Hi @morcef!

This update has merged to main and released as part of [v1.0.4](https://github.com/CrowdStrike/falconpy/releases/tag/v1.0.4).

**Reply by morcef** [reply]
Hi @jshcodes :)

I updated the package, and that removed the warning!
Thank you!
